pkg/fileutils/fileutils.go

```diff
@@ -55,19 +55,24 @@ func NewPatternMatcher(patterns []string) (*PatternMatcher, error) {
 	return pm, nil
 }
 
-// Matches matches path against all the patterns. Matches is not safe to be
-// called concurrently
+// Matches returns true if "file" matches any of the patterns
```

Comment:
Could we rename this and remove the `range`  loop. So all callers need to update and choose if they don't need to match parents or if they track the parents on the caller side.

Reply:
What's the use case for not matching parent dirs? I think that would give incorrect behavior with a number of patterns. Are there cases where we know in advance we wouldn't encounter such patterns?

Removing `Matches` would complicate code that isn't doing a walk. For example, `removeDockerfile` currently uses `Matches`, and it would need similar logic to the loop in `Matches`. Even `TarWithOptions`, which I modified to use `MatchesUsingParentResult`, uses `Matches` as a convenience on the first file/dir it encounters matching the current `include` prefix.

Reply:
I think it is unexpected that this function now has `m*n` complexity and this would make it sure it doesn't go unnoticed. If you think there are cases that need parent matches but would still prefer to use `Matches` (eg. they only call a single file anyway) maybe leave it as is but rename to `MatchesOrAnyParentMatches`.

Reply:
Renamed it. I kept the original (buggy) implementation as well because some vendored code calls it. We can remove it as a second step.